packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -3027,10 +3073,42 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         type: ContainerMessageType,
         contents: any,
         batch: boolean,
+        contentLength: number,
+        serializedContent?: string,
         appData?: any,
     ) {
         this.verifyNotClosed();
         assert(this.connected, 0x259 /* "Container disconnected when trying to submit system message" */);
+
+        if (this.runtimeOptions.compressionOptions?.minimumSize !== undefined &&
+            serializedContent &&
+            contentLength > this.runtimeOptions.compressionOptions.minimumSize
+        ) {
+            this.compressedOpCount++;
+
+            const compressionStart = Date.now();
+            const contentsAsBuffer = new TextEncoder().encode(serializedContent);
+            const compressedContents = compress(contentsAsBuffer);
+            const compressedContent = IsoBuffer.from(compressedContents).toString("base64");
+            const duration = Date.now() - compressionStart;
+
+            if (this.compressedOpCount % 100) {
+                this.mc.logger.sendPerformanceEvent({
+                    eventName: "compressedOp",
+                    duration,
+                    sizeBeforeCompression: Math.ceil(contentLength / 3.0) * 4, // Cost of base64 encoding
+                    sizeAfterCompression: compressedContent.length,
+                });
+            }
+
+            const compressedPayload: ContainerRuntimeMessage = { type, contents: compressedContent };
```

Comment:
It might be beneficial to not use compressed payload if compressedContent.length >= contentLength. Not sure how often it happens, but it feels like worth the trouble to add this optimization (though it adds extra code path that is harder to test)

Reply:
Added a block to only submit the compressed payload if it's actually smaller and also included an e2e test for each path